onnxruntime/core/providers/webnn/builders/impl/normalization_op_builder.cc

```diff
@@ -114,56 +76,26 @@ Status NormalizationOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder
     ORT_RETURN_IF_NOT(input_defs.size() == 5, "BatchNormalization requires five inputs.");
     emscripten::val mean = model_builder.GetOperand(input_defs[3]->Name());
     emscripten::val variance = model_builder.GetOperand(input_defs[4]->Name());
-    // Enlarge 1-D mean and variance to new scale shape.
-    emscripten::val reshape_mean =
-        model_builder.GetBuilder().call<emscripten::val>("reshape", mean, emscripten::val::array(new_scale_shape));
-    emscripten::val reshape_variance =
-        model_builder.GetBuilder().call<emscripten::val>("reshape", variance, emscripten::val::array(new_scale_shape));
-
-    std::vector<uint32_t> axes = {0};
-    for (uint32_t i = 2; i < rank; i++) {
-      axes.push_back(i);
+    if (model_builder.GetPreferredLayout() == DataLayout::NHWC) {
+      options.set("axis", rank - 1);
     }
-
-    options.set("axes", emscripten::val::array(axes));
-    options.set("mean", reshape_mean);
-    options.set("variance", reshape_variance);
-    output = model_builder.GetBuilder().call<emscripten::val>("meanVarianceNormalization", input, options);
+    output = model_builder.GetBuilder().call<emscripten::val>("batchNormalization", input, mean, variance, options);
   } else if (op_type == "LayerNormalization") {
     int64_t axis = helper.Get("axis", -1);
     axis = HandleNegativeAxis(axis, rank);
     std::vector<uint32_t> axes(rank - SafeInt<uint32_t>(axis));
-    std::iota(axes.begin(), axes.end(), axis);
+    if (model_builder.GetPreferredLayout() == DataLayout::NHWC) {
+      std::iota(axes.begin(), axes.end(), axis - 1);
```

Comment:
Does this properly handle cases of NWC and NHWDC?

What about 4D with axis = 1 (or really *firstAxis* based on how it works in ONNX)?
For NCHW, it would produce axes = [1,2,3], but for NCHW, it would yield [0,1,2] when you'd still want the last three dimensions with axes = [1,2,3].

The firstAxis = 2 case is interesting too, because the standard NCHW layout would yield axes = [2,3], but for NCHW layout, you'd need axes = [1,3].

Reply:
Maybe the key is whether the axis affects channel dimension? Is it correct to use cond `model_builder.GetPreferredLayout() == DataLayout::NHWC && axis > 1` ?

Reply:
(I'll have to think about it tomorrow - 1am here now :b...)

Reply:
>  Is it correct to use cond ... axis > 1

Yes, that appears to be the key, that if axis > 1, then start the iota - 1:

3D
axis = 0, NCW = [0,1,2], NWC = [0,1,2]
axis = 1, NCW = [1,2], NWC = [1,2]
axis = 2, NCW = [2], NWC = [1]

4D
axis = 0, NCHW = [0,1,2,3], NHWC = [0,1,2,3]
axis = 1, NCHW = [1,2,3], NHWC = [1,2,3]
axis = 2, NCHW = [2,3], NHWC = [1,2]
axis = 3, NCHW = [3], NHWC = [2]

5D
axis = 0, NCDHW = [0,1,2,3,4], NDHWC = [0,1,2,3,4]
axis = 1, NCDHW = [1,2,3,4], NDHWC = [1,2,3,4]
axis = 2, NCDHW = [2,3,4], NDHWC = [1,2,3]
axis = 3, NCDHW = [3,4], NDHWC = [2,3]
axis = 4, NCDHW = [4], NDHWC = [3]